File: src/main/java/io/github/zenwave360/zdl/antlr/ZdlModelPostProcessor.java
package io.github.zenwave360.zdl.antlr;

import io.github.zenwave360.zdl.antlr.JSONPath;
import io.github.zenwave360.zdl.antlr.ZdlModel;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class ZdlModelPostProcessor {

    public static ZdlModel postProcess(ZdlModel model) {
        var entities = model.getEntities();
        var inputs = model.getInputs();
        var outputs = model.getOutputs();
        var enums = model.getEnums();
        var events = model.getEvents();

        var fields = JSONPath.get(model, "$..fields[*]", List.<Map<String, Object>>of());
        for (var field : fields) {
            if(entities != null && entities.containsKey(field.get("type"))) {
                field.put("isEntity", true);
                field.put("isComplexType", true);
            }
            if(enums != null && enums.containsKey(field.get("type"))) {
                field.put("isEnum", true);
                field.put("isComplexType", true);
            }
            if(inputs != null && inputs.containsKey(field.get("type"))) {
                field.put("isInput", true);
                field.put("isComplexType", true);
            }
            if(outputs != null && outputs.containsKey(field.get("type"))) {
                field.put("isOutput", true);
                field.put("isComplexType", true);
            }
            if(events != null && events.containsKey(field.get("type"))) {
                field.put("isEvent", true);
                field.put("isComplexType", true);
            }
        }

        var allEntitiesAndEnums = new HashMap<>();
        if(entities != null) {
            allEntitiesAndEnums.putAll(entities);
        }
        if(enums != null) {
            allEntitiesAndEnums.putAll(enums);
        }
        if(inputs != null) {
            allEntitiesAndEnums.putAll(inputs);
        }
        if(outputs != null) {
            allEntitiesAndEnums.putAll(outputs);
        }
        model.put("allEntitiesAndEnums", allEntitiesAndEnums);

        return model;
    }
}

File: src/commonMain/kotlin/io/zenwave360/zdl/antlr/ZdlModelPostProcessor.kt
package io.zenwave360.zdl.antlr

class ZdlModelPostProcessor {
    companion object {
        fun postProcess(model: ZdlModel): ZdlModel {
            val aggregates = model.getAggregates()
            val entities = model.getEntities()
            val inputs = model.getInputs()
            val outputs = model.getOutputs()
            val enums = model.getEnums()
            val events = model.getEvents()

            val fields = JSONPath.get(model, "$..fields[*]", listOf<MutableMap<String, Any?>>())
            for (field in fields) {
                val type = field["type"]
                if (type != null) {
                    field["isComplexType"] = false
                    if (entities.containsKey(type)) {
                        field["isEntity"] = true
                        field["isComplexType"] = true
                    }
                    if (enums.containsKey(type)) {
                        field["isEnum"] = true
                        field["isComplexType"] = true
                    }
                    if (inputs.containsKey(type)) {
                        field["isInput"] = true
                        field["isComplexType"] = true
                    }
                    if (outputs.containsKey(type)) {
                        field["isOutput"] = true
                        field["isComplexType"] = true
                    }
                    if (events.containsKey(type)) {
                        field["isEvent"] = true
                        field["isComplexType"] = true
                    }
                }
            }

            val allEntitiesAndEnums = mutableMapOf<String, Any?>()
            allEntitiesAndEnums.putAll(aggregates)
            allEntitiesAndEnums.putAll(entities)
            allEntitiesAndEnums.putAll(enums)
            allEntitiesAndEnums.putAll(inputs)
            allEntitiesAndEnums.putAll(outputs)
            model["allEntitiesAndEnums"] = allEntitiesAndEnums

            return model
        }
    }
}

